### DISCUSSION TITLE
Best way to increase performance with MongoDB for large number of recrods?

### DISCUSSION THREAD

**Discussion by MartinDawson** [cloud-carbon-footprint/cloud-carbon-footprint#1240] [Q&A] [answered]
We have the UI (same as your demo UI) and API deployed on our own Azure instance and we use MongoDB caching with a MongoDB azure instance database.

The issue is, that when the UI does the carbon emission API calls, it takes ages to resolve (30~ seconds) for just ONE day of data.

I.e these api calls:

![MicrosoftTeams-image (8)](https://github.com/cloud-carbon-footprint/cloud-carbon-footprint/assets/15030491/323e8d1d-b76c-4b6d-bb6f-f28de84b195e)

take forever to resolve. This seems to be because the code is doing this mongoDB call:

`database.listCollections` which takes a long time to resolve because there are thousands of services to return for each day:

![MicrosoftTeams-image (10)](https://github.com/cloud-carbon-footprint/cloud-carbon-footprint/assets/15030491/9a22460d-8b57-4051-92c2-961b4b713d02)

In our mongoDB instance we have 350,000 records stored in the database under the `estimates-by-day` document grouping.

This is just for the past 3 months as well.

![MicrosoftTeams-image (9)](https://github.com/cloud-carbon-footprint/cloud-carbon-footprint/assets/15030491/903bf288-ccb1-4be9-878a-90f252c6995f)

So it seems that this is the reason it's extremely slow, it has to loop over 350,000 records.

How can we make this faster?

The only way I can think of is that we have our own database and API endpoints set up this new backend and database specifically for our UI and we pre-store the azure data in the database by date for each subscription before hand.

Any other ideas on this?

**Comment by camcash17** [comment, marked as answer]
Hello @MartinDawson, thanks for raising this!

Currently, it is looking like you are making the request for about a month. I am curious how you are able to tell that it is around 30 seconds per day.

Also, you may want to check out in our [configurations glossary](https://www.cloudcarbonfootprint.org/docs/configurations-glossary#client-package---all-variables-are-optional) the option to set the pagination number. This may help with the performance if you decrease it.

Another suggestion is to use the `seed-cache-file` script. This will allow you to fetch data more as a background job, so when it comes time to load the UI, the data is already cached. You may also want to just to do this cache historical data, and implement a daily request system so they you never need to fetch more than a days worth of data each time.

Hope this helps!
